File: 02_two_pointers/01_is_palindrome.py
def is_palindrome(s: str) -> bool:
    s = s.lower()
    i = 0
    j = len(s) - 1
    while i < j:
        if i >= len(s) or j < 0:
            return True
        if not s[i].isalnum():
            i += 1
            continue
        if not s[j].isalnum():
            j -= 1
            continue
        if not s[i] == s[j]:
            return False
        i += 1
        j -= 1
    return True


def test_is_palindrome():
    assert is_palindrome("racecar")
    assert not is_palindrome("hello")
    assert not is_palindrome("tab a cat")

if __name__ == "__main__":
    test_is_palindrome()

File: 02_two_pointers/01_is_palindrome.cpp
#include <cassert>
#include <iostream>
#include <string>
#include <cctype>

using std::string;

bool is_palindrome(string s)
{
	for (char& c : s) {
		c = std::tolower(c);;
	}

	int i = 0;
	int j = s.length() - 1;

	while (i < j) {
		if (i >= s.length() || j < 0) {
			return true;
		}

		while (!(s[i] >= 'a' && s[i] <= 'z') && i < j) i++;
		while (!(s[j] >= 'a' && s[j] <= 'z') && i < j) j--;

		if (!(s[i] == s[j])) {
			return false;
		}

		i++;
		j--;
	}

	return true;
}

void test_is_palindrome()
{
	assert(is_palindrome("racecar"));
	assert(!is_palindrome("hello"));
	assert(!is_palindrome("tab a cat"));

	std::cout << "All tests passed" << std::endl;
}

int main()
{
	test_is_palindrome();
	return 0;
}
